src/forecasts.jl

```diff
@@ -1,316 +1,91 @@
-        add_forecast!(forecasts, forecast)
-    end
-
-    return forecasts
-end
-
-function Base.length(forecast::Forecast)
-    return get_horizon(forecast)
-end
-
-function Base.getindex(forecast::Forecast, args...)
-    return _split_forecast(forecast, getindex(get_data(forecast), args...))
-end
-
-Base.first(forecast::Forecast) = head(forecast, 1)
-
-Base.last(forecast::Forecast) = tail(forecast, 1)
-
-Base.firstindex(forecast::Forecast) = firstindex(get_data(forecast))
-
-Base.lastindex(forecast::Forecast) = lastindex(get_data(forecast))
-
-Base.lastindex(forecast::Forecast, d) = lastindex(get_data(forecast), d)
-
-Base.eachindex(forecast::Forecast) = eachindex(get_data(forecast))
-
-Base.iterate(forecast::Forecast, n = 1) = iterate(get_data(forecast), n)
-
 """
-Refer to TimeSeries.when(). Underlying data is copied.
+Return the count of forecast windows.
 """
-function when(forecast::Forecast, period::Function, t::Integer)
-    new = _split_forecast(forecast, TimeSeries.when(get_data(forecast), period, t))
-
+function get_count(forecast::Forecast)
+    return length(get_data(forecast))
 end
 
 """
-Return a forecast truncated starting with timestamp.
+Return the forecast interval as a Dates.Period.
 """
-function from(forecast::T, timestamp) where {T <: Forecast}
-    return T(get_label(forecast), TimeSeries.from(get_data(forecast), timestamp))
+function get_interval(forecast::Forecast)
+    k = keys(get_data(forecast))
+    if length(k) == 1
+        return Dates.Second(0)
+    end
+    first_key, state = iterate(k)
+    second_key, state = iterate(k, state)
+    return second_key - first_key
 end
 
 """
-Return a forecast truncated after timestamp.
+Return the forecast window corresponsing to initial_time.
 """
-function to(forecast::T, timestamp) where {T <: Forecast}
-    return T(get_label(forecast), TimeSeries.to(get_data(forecast), timestamp))
+function get_window(forecast::Forecast, initial_time::Dates.DateTime)
+    return TimeSeries.TimeArray(
+        make_timestamps(forecast, initial_time),
+        forecast.data[initial_time],
+    )
 end
 
 """
-Return a forecast with only the first num values.
+Return the forecast window corresponsing to interval index.
 """
-function head(forecast::Forecast)
-    return _split_forecast(forecast, TimeSeries.head(get_data(forecast)))
-end
-
-function head(forecast::Forecast, num)
-    return _split_forecast(forecast, TimeSeries.head(get_data(forecast), num))
+function get_window(forecast::Forecast, index::Int)
+    return get_window(forecast, index_to_initial_time(forecast, index))
 end
 
 """
-Return a forecast with only the ending num values.
+Iterate over all forecast windows.
 """
-function tail(forecast::Forecast)
-    return _split_forecast(forecast, TimeSeries.tail(get_data(forecast)))
-end
-
-function tail(forecast::Forecast, num)
-    return _split_forecast(forecast, TimeSeries.tail(get_data(forecast), num))
+function iterate_windows(forecast::Forecast)
```

Comment:
Do we need a method of this function implemented on `Component`?  This method will only work on data that has been loaded into memory.  For example, suppose the user calls `get_time_series` with `count = 10` but there are 100,000 windows.  This function will only iterate over the 10 windows.  Users might want a function that iterates over all windows while only keeping one window in memory at a time.  It would slow because there would be many small disk reads, but it would use very little memory.

Reply:
This has been solved by `ForecastCache`.